ath79: add LED for Ubiquiti airCube

c-mauderer opened the pull request:
This adds a simple LED driver for the single LED of the Ubiquiti
airCube. The kernel patches have been pulled from linux-next on tag
next-2019-05-24.

Signed-off-by: Christian Mauderer <oss@c-mauderer.de>

c-mauderer commented:
Note that this is the second try. The first one has been #2029.

ynezz commented on target/linux/ath79/modules.mk:
```diff
@@ -13,4 +13,19 @@ define KernelPackage/leds-reset/description
+define KernelPackage/leds-spi-byte
+  SUBMENU:=$(LEDS_MENU)
+  TITLE:=simple SPI LED controller support
+  DEPENDS:= @TARGET_ath79
```
Do you've any reason for this dependency?

c-mauderer replied:
Currently it only works on the airCube and there is no other device supporting it. But you are right: it's not really necessary. I'll remove it.

c-mauderer replied:
Oh and maybe its a copy and paste one: leds-reset in the same file has it too.

c-mauderer replied:
Just noted: With the change I have to move the patches to the common kernel patches and rebuild everything. Please wait with the merge till I did that.

ynezz commented:
> The kernel patches have been pulled from linux-next on tag
> next-2019-05-24.

Excellent work, thanks!

c-mauderer committed:
kernel: add leds-spi-bytes module

This adds a package for leds-spi-bytes. The kernel patches have been
pulled from linux-next on tag next-2019-05-24.

Note that the patches don't work unchanged on 4.9 so they haven't been
ported there.

Signed-off-by: Christian Mauderer <oss@c-mauderer.de>

c-mauderer commented:
> > The kernel patches have been pulled from linux-next on tag
> > next-2019-05-24.
> 
> Excellent work, thanks!

Bringing it into the kernel has changed the driver quite a bit but it was a nice experience. So thanks for leading me there.

c-mauderer committed:
ath79: add LED for Ubiquiti airCube

This adds LED support for the single LED of the Ubiquiti airCube.

Signed-off-by: Christian Mauderer <oss@c-mauderer.de>

c-mauderer commented:
I moved the module to be a generic kernel module. I didn't find an option to compile all targets (is there one?) so I picked some random samples:

Tested and running:
- airCube ISP
- airCube ISP (modified to use a 4.19 Kernel)

At least building without build problems:
- TI AR7
- Freescale i.MX6
- Samsung s3c24xx/s5pv210

The patches don't apply on 4.9. So I added a dependency for the kernel to _not_ be 4.9. I hope that is OK. The alternative would be to backport the patches. But it seems that 4.9 is a barely used version anyway.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

ynezz commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -199,6 +199,9 @@ ubnt,nanostation-ac)
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "ubnt:blue:rssi3" "wlan0" "76" "100"
 	;;
+ubnt,acb-isp)
+	ucidef_set_led_default "system" "SYSTEM" "white:status" "1"
```
Sorry, but while going through the changes here locally again, I've noticed, that I've missed following previously.

1. The name of the LED should follow boardname(or vendor):color:function, so it should be `ubnt:white:status`
2. This line is superfluous, as you're doing already the same in DTS

c-mauderer replied:
1. OK. I can change that.
2. Does that mean that I should just completely remove the "ucidef_set_led_default" line?

adrianschmutzler replied:
2. Yes. Note that there is not a single ucidef_set_led_default in 01_leds for ath79.

c-mauderer replied:
Hello @adrianschmutzler,
thanks for the answer. I already discussed that with ynezz a bit farther down. Currently I still have the problem that the SPI is too fast on the device. I started a discussion (and didn't receive a reply for two weeks until I pinged the subsystem maintainer) on the Linux SPI mailing list. The Maintainer suggested a rough direction how I could slow down the driver without having too much performance impact on other users. I wanted to take a look at that (most likely) this weekend. If I have something that is acceptable for Linux, I'll update the patch set here - including the changes already suggested by ynezz.

ynezz commented on target/linux/ath79/dts/qca9533_ubnt_acb-isp.dts:
```diff
@@ -20,6 +20,28 @@
+			spi-max-frequency = <100000>;
+
+			led {
+				label = "white:status";
```
The name of the LED should follow boardname(or vendor):color:function, so it should be `ubnt:white:status`

c-mauderer replied:
Again: I'll change it.

ynezz commented:
Now that you've the LED functional, could you please consider adding LEDs aliases as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 4. If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dts#L13)) https://github.com/openwrt/openwrt/blob/fe7d965ea95e78905328fe5425c8e90e3bf11e58/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dts#L13-L18

c-mauderer commented:
I'll have a look at the checklist. Should I create a new pull request with the changes? I don't think I can re-open that one, right?

ynezz commented:
I can reopen it if you want to do the changes by yourself or you can just tell me if this changes are fine with you https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=5c7e1dd92e71a356c8ac529c97ca02f5f9d58b44

ynezz commented:
Ideally of course test it again on your device as I'm not sure if everything works as expected with the SPI LEDs, although I'm 99.9% sure it should work.

ynezz commented:
I've just added [another change](https://github.com/ynezz/openwrt/blob/a3f8def648fc9068d111532b436673162bb9743e/target/linux/ath79/dts/qca9533_ubnt_acb-isp.dts#L36-L43) based on this [discussion](https://github.com/openwrt/openwrt/pull/2056#discussion_r285335699).

c-mauderer commented:
I tried to build your branch and it works fine for a 4.14 as well as a 4.19 kernel. The changes to the patches are looking good so no need to re-open this pull-request.

The only difference I've found so far is that now there is no longer a LED pre-configured in luci. But it can be added manually if it should show something other than just "on".

Regarding the aliases: Should I see a difference somewhere? In other words: Should the LED now do something special like for example blink during failsafe?

ynezz commented:
> I tried to build your branch and it works fine for a 4.14 as well as a 4.19 kernel. The changes to the patches are looking good so no need to re-open this pull-request.

Ok, thanks.

> The only difference I've found so far is that now there is no longer a LED pre-configured in luci. But it can be added manually if it should show something other than just "on".

Maybe this could be fixed somehow in LuCI.

> Regarding the aliases: Should I see a difference somewhere? In other words: Should the LED now do something special like for example blink during failsafe?

Indeed, you should spot the difference during the normal boot as well, the LED should blink. All the magic is in generic [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh).

c-mauderer commented:
> > Regarding the aliases: Should I see a difference somewhere? In other words: Should the LED now do something special like for example blink during failsafe?
> 
> Indeed, you should spot the difference during the normal boot as well, the LED should blink. All the magic is in generic [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh).

I had a quick look at the LED and it seems that timber based blinking isn't really working correctly. It seems that the LED only blinks if there is some other system activity (like typing something on the serial console). Maybe I missed some system timer configuration ...

I'll have to have a more detailed look at that. But most likely I can't do much before Thursday due to lack of time.

ynezz commented:
> I had a quick look at the LED and it seems that timber based blinking isn't really working correctly.

Ok, thanks for testing. That generic diags.sh based solution is using standard kernel LED timer trigger, so if it doesn't work properly, then it might be some issue.

> It seems that the LED only blinks if there is some other system activity (like typing something on the serial console). Maybe I missed some system timer configuration ...

I would probably start with SPI, just verify, that the LED (or the IC driving it) is getting the data. Maybe it's CS/SCK interfering somehow?

> I'll have to have a more detailed look at that. But most likely I can't do much before Thursday due to lack of time.

There is no need to hurry with anything, I've this still only in my staging tree, I didn't pushed it yet into master as I would like to run it through some build tests before I would push it there.

So how would you like to proceed? Should I reopen the PR so you can continue working on this?

c-mauderer commented:
> > I had a quick look at the LED and it seems that timber based blinking isn't really working correctly.
> 
> Ok, thanks for testing. That generic diags.sh based solution is using standard kernel LED timer trigger, so if it doesn't work properly, then it might be some issue.
> 
> > It seems that the LED only blinks if there is some other system activity (like typing something on the serial console). Maybe I missed some system timer configuration ...
> 
> I would probably start with SPI, just verify, that the LED (or the IC driving it) is getting the data. Maybe it's CS/SCK interfering somehow?
> 

Thanks for the hint. I'll try to take a look at that.

The other thing that I want to try is adding a GPIO LED to some pin and have a look at the behaviour of that LED. I hope I can find an unused one. Although it's a fine pitch I think I can solder a wire to the outer row  of pins.

> > I'll have to have a more detailed look at that. But most likely I can't do much before Thursday due to lack of time.
> 
> There is no need to hurry with anything, I've this still only in my staging tree, I didn't pushed it yet into master as I would like to run it through some build tests before I would push it there.
> 
> So how would you like to proceed? Should I reopen the PR so you can continue working on this?

Maybe re-opening the PR would be a good idea. Sorry for wasting your time. I missed the behaviour during blink. I only tested on, off, some triggers like network traffic and different brightness levels (the last one via the /sys/class/led interface). With those I missed that a normal blink doesn't work correctly.

c-mauderer commented:
I at least found out what's wrong: The SPI frequency is much to high. Instead of about 100kHz, I get about 2.3 to 2.5 MHz. But I haven't found a reason yet why the frequency is that high. It just seems that the spi-gpio ignores the spi-max-freqency of the child. Did you have seen some similar effect on some other target?

c-mauderer commented:
Most likely this will be a problem that needs some more time. Currently the GPIO-SPI in Linux just `clocks "as fast as we can"`: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/spi/spi-gpio.c?h=v5.2-rc2#n120

I expect that it will be quite hard to find some solution that is fast enough in the normal case that it will be accepted as a replacement for that macro. I'll have to start searching through the history of the mailing list that is responsible for the SPI part and if there is no discussion how that could be solved, I'll have to start a discussion.

With that in mind: You most likely won't hear anything new from me in the next few weeks.

adrianschmutzler commented:
Although it is not related to the aim of this PR, I wonder whether you could help me with some information about your device(s):

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.

c-mauderer commented:
> Although it is not related to the aim of this PR, I wonder whether you could help me with some information about your device(s):

You could have mentioned me in your PR with a @c-mauderer. As far as I know github I should have sent a notification. I also just added my mail address to my github profile. So if you have any more questions, you can reach me by that.

> 
> I'm currently working on a PR to store the label-mac-address in the DTS (#2159).
> 
> I would appreciate if you could provide information about your device:

The MAC of `/sys/firmware/devicetree/base/ahb/apb/wmac@18100000/mac-address` matches the label. The other ones add either a `00:00:00:01:00:00` for `eth@19000000` (eth1) or `02:00:00:01:00:00` for `eth@1a000000` (eth0).

adrianschmutzler commented:
Thanks, so this one will be a `label-mac-device = &wmac;`

c-mauderer commented:
> What's the current status here? Is this still being worked on or just left open as a reminder?

I had a discussion with two of the Linux SPI maintainers what would be a good way to implement a slowed down GPIO-SPI here: https://marc.info/?t=156018626500001&r=1&w=2

I wanted to write a prototype for a patch. I always plan to do it sometime but somehow I didn't manage to find the time and desire to do it yet. So it's still open. If it would be better to close the pull request to keep the number of open pull requests small I won't object. I can always re-open a new one as soon as I manage to get it done.

adrianschmutzler commented:
@c-mauderer 

Personally, I'd say that a PR is not a bug report, so if the current state cannot be merged and you do not plan to change that status in the foreseeable future, it would be better to close the PR and reopen it when a new patchset is available. (I'd prefer to reopen the existing PR instead of creating a new one, so history is preserved).

BUT,
1. I've just learned that a PR cannot be reopened if it was closed by an admin. It can be opened again by the author if it was closed by the author before, though.
2. This was reopened by @ynezz (though on your request) and I do not want to interfere with his judgement, especially since I've not been following the discussion.

ynezz commented:
> If it would be better to close the pull request to keep the number of open pull requests small I won't object. I can always re-open a new one as soon as I manage to get it done.

Sounds like a plan, thanks.

c-mauderer commented:
Like discussed I'll close the PR for now and re-open if I continue to work on it.

NextBlaubeere commented:
Hi everyone. What is the state about this currently?
OpenWRT on AirCube AC is awesome, but the LED sucks.

stklcode commented:
I gave it a short try couple of months ago on 21.02 basis (https://github.com/stklcode/openwrt/tree/aircube-led-21.02)

The SPI LED driver is in 5.4 kernel, so I omitted the backports required for 4.9 and added the kmod-leds-spi-byte package.
Think we still need a delay patch (there are a few implementations around the web) to throttle the SPI frequency.
And last but not least the devicetree update.

The branch linked above does compile,  hat unfortunately the LED is still not working as expected. Didn’t find the time for further investigation yet. I’ll probably give it another try soon. Any hints appreciated.

c-mauderer commented:
That topic is still open. I never finished the patch for the Linux kernel to slow down the SPI. My current ideas would either slow down all targets (which wouldn't be acceptable) or would need some tricky pre-processor magic together with switching the functions (which is a complex solution and therefore I didn't finish it).

Another possibility would be to add a Linux Kernel patch to OpenWRT that replaces the "spidelay" with some function that really sleeps: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/spi/spi-gpio.c?h=v5.2-rc2#n120
Problem with that is that it would be a device specific patches that would never go into the Linux kernel which adds maintenance load to OpenWRT.

c-mauderer commented:
PS: @NextBlaubeere: Please mark it clearly if you cross-post on multiple locations. It's simpler for someone who finds one answer to find the other too. So in this case a link to the forum would have been nice: https://forum.openwrt.org/t/add-openwrt-support-for-ubiquity-aircube-isp-ubnt-acb-isp/32999/111

NextBlaubeere commented:
I am sorry. I was not fully aware of the netiquette about cross posting. I will do in future. 👍 

It took me a long time, but i finally had some time to dig into this topic and I realised this is over my head.

Instead of getting the patch into the Linux kernel the proper way, couldn't we just do it local? So we could compile a test-build for our device, with the ability to turn off the LED, even though it is not supported or maintenanced in future.
I really do not care about it being futureproof, i just want the LED to stop blinking.


> I only tested on, off (...)

this suggest you were indeed able to compile a version with the functionality to turn the led off. If you have it still laying around, would you mind sharing it?



> I gave it a short try couple of months ago on 21.02 basis (https://github.com/stklcode/openwrt/tree/aircube-led-21.02)
> 
> The SPI LED driver is in 5.4 kernel, so I omitted the backports required for 4.9 and added the kmod-leds-spi-byte package. Think we still need a delay patch (there are a few implementations around the web) to throttle the SPI frequency. And last but not least the devicetree update.
> 
> The branch linked above does compile, but unfortunately the LED is still not working as expected. Didn’t find the time for further investigation yet. I’ll probably give it another try soon. Any hints appreciated.

Thank you for your efforts trying to compile a build. When you say "the LED is still not working as expected", does it mean controlling it does not work at all, or is the functionality just limited? I would really just like to turn it off.

NextBlaubeere commented:
> I gave it a short try couple of months ago on 21.02 basis (https://github.com/stklcode/openwrt/tree/aircube-led-21.02)

Hey, I just compiled and tried your OpenWRT code on my AirCube AC. The image does work, but the LED-control does unfortunately not. Too bad it is so damn hard to turn it off.
Let me know, if I can do something for you guys, to improve the issue.

stklcode commented:
Same for the 23.03 port. Right now I’m not even sure that the DTS is correct for the AC model and I don’t own an ISP variant, so I can’t tell if the driver doesn’t do the job or the pinout is simply wrong. Didn’t find the time to teardown a device and check it at hardware level.

stklcode commented:
Heureka!
It's finally done, full LED control on my Aircube AC.

It's 3 commits on top of the 22.03 or 21.02 branch: https://github.com/stklcode/openwrt/tree/aircube-led-21.02
https://github.com/stklcode/openwrt/tree/aircube-led-22.03
(only tested the 22.03 so far)

The delay has to be enabled in the `make kernel_menuconfig` step:
Device Drivers -> SPI support -> Enable delays in the GPIO-based bitbanging SPI Master

The driver by @c-mauderer which is now upstream works fine. The delay patch does the rest. The LED(-cluster) is dimmable from 0 to 100% with
`# echo 0 > /sys/class/leds/ubnt:white:status/brightness`
to
`# echo 63 > /sys/class/leds/ubnt:white:status/brightness`

I did some reverse engineering on the hardware. It's the very same Sonix SPI controller, however the GPIO pinout is different from the ISP model (different SoC, why should it be the same)

<details>
  <summary>pinout image</summary>

  ![AirCubeAC_LED_pinout](https://user-images.githubusercontent.com/12963621/163589391-23818442-aded-44de-bfe3-fe5073773ee7.jpg)
</details>

c-mauderer commented:
Hello @stklcode and @NextBlaubeere
sorry that I didn't reacted earlier. I was quite busy and also I planned to throw something together on some weekend, I just somehow didn't manage to find the time.
@stklcode: That sounds great. I would have to take a look at my first attempt to create a kernel patch but I think I tried something similar and the Linux Kernel maintainers didn't like it. But maybe you have more luck so don't hesitate to try to send it upstream. Alternatively you have to talk to @ynezz whether a pull request for OpenWRT might could have success or whether the device specific kernel patch is a blocker.

ynezz commented:
> whether the device specific kernel patch is a blocker.

Device specific patches doesn't scale well. Indeed, ideally we want upstream solution, in this case it looks like some boolean DTS knob `per-bit-slowdown` or such, which would enable those per bit delays could be a better fit:

```dts
led_spi {
		compatible = "spi-gpio";
                per-bit-slowdown;
};
```

c-mauderer commented:
@ynezz: Thanks for the answer to that. Enabling the delay with a DTS knob means that there has to be some "if" in the spi-gpio code. The problem is that this is a highly optimized location. Adding even a single "if" makes the maintainers of the spi-gpio code quite unhappy.

From what I have analysed (quoting from a few years old memory so I'm maybe not 100% accurate) it would be possible to create two sets of functions that do the low level pin toggling. One set with a proper delay function that adds a delay depending on the speed setting in the dts and one set that just works as fast as possible (that's what the code does at the moment). Somewhere a function pointer is used anyway so that shouldn't slow down the bus. The problem is that this needs either code duplication (which is very likely not acceptable) or some hacky pre-processor magic. The pre-processor magic is a bit tricky to write in a way that the code is still readable. That's where I gave up in 2019.

NextBlaubeere commented:
> Heureka! It's finally done, full LED control on my Aircube AC.
> 
> It's 3 commits on top of the 22.03 or 21.02 branch: https://github.com/stklcode/openwrt/tree/aircube-led-21.02 https://github.com/stklcode/openwrt/tree/aircube-led-22.03 (only tested the 22.03 so far)
> 
> The delay has to be enabled in the `make kernel_menuconfig` step: Device Drivers -> SPI support -> Enable delays in the GPIO-based bitbanging SPI Master
> 
> The driver by @c-mauderer which is now upstream works fine. The delay patch does the rest. The LED(-cluster) is dimmable from 0 to 100% with `# echo 0 > /sys/class/leds/ubnt:white:status/brightness` to `# echo 63 > /sys/class/leds/ubnt:white:status/brightness`
> 
> I did some reverse engineering on the hardware. It's the very same Sonix SPI controller, however the GPIO pinout is different from the ISP model (different SoC, why should it be the same)
> pinout image

Halleluja! Thank you so much guys for your effort.
I tested both the 21.02 and 22.03 and each build does a great job of having the LED controlable. It would be really nice to have this in the official openWRT repository, but you make me very happy with this build alone. I will probably use it for years to come.
Thank you!

stklcode commented:
From what I recall there a couple of possible solutions für a delay-patch. Discusion dates at least to the 2.6 days.

1. Config-level patch with macro conditional as proposed in "my" solution.
  Pros: Does not interfere with existing setups, as decision is made at preprocessor level. It's a quite minimalistic change.
  Cons: Was rejected 12 years ago. Main argument was, this should be a device-level decision, not config-level, because there might be systems with multiple SPI buses of which one requires full speed.
  
  Personally, I don't follow this con-argumentation. I'm not aware of any system with multiple GPIO-SPI buses that require different configuration. If throttling is required for either one, chances are high that a higher clock rate is acceptable for the other, too. This case should be really rare. So far this doe not influence hardware SPI drivers, so a mix of hard- and soft-buses is still fine.

2. Per-device DTS configuration.
  Pros: mixed-systems are possible with a single driver.
  Cons: Requires at least one inline conditional which slows down every driver, delayed or not.
  
  I can see the impact here. There are "slow" embedded systems where not only LED-controllers are driven by SPI and a couple of CPU cycles do have impacts.

3. Duplicate the GPIO-SPI driver, one as is, one with delay.
  Pros: Per-device decision without interference on undelayed buses.
  Cons: Degraded maintainability of duplicated code. Extended binary size (~2x) for both modules.
  
  I'm no friend of duplications and so are most maintainers.

4. Some hacky preprocessor magic to find a solution inbetween.
  The description tells the story I guess. Code is hard to write, hard to read, hard to maintain and not that many systems take profit from it.
  Chances are high that a quality patch of this kind might be accepted, but it's not trivial at all.

5. Rewrite the GPIO-SPI-LED driver, s.t. it no longer relies on the GPIO-SPI layer.
   The driver is a specialized piece of code anyway. But this adds quite a few duplications, too. I have my doubts that such solution really is something someone wants to maintain.

----

I still prefer the first. I also see that maintaining even a small patch forever is overhead we want to avoid. I'll probably try to (re)start or join a discussion upstream and maybe get a simpler solution in place.

NextBlaubeere commented:
This has never made it into the official OpenWRT repository, has it?

c-mauderer commented:
> This wonderful patch for the airCubes LED controller has never made it into the official OpenWRT repository, has it?

Correct. I never finished the patch. Basically the big problem was (and still is) that the Linux GPIO SPI driver is optimized for speed. On the chip used in the airCube it is too successful. The small controller that controls the LED can't keep up with the SPI and looses some bits.

It would be simple to create a hacked patch, that just adds some delay in the Linux GPIO SPI (it's more or less prepared in the kernel driver) but that won't be acceptable as a patch for the general Linux kernel because it slows down the GPIO SPI for all systems using it. Difficult thing is: The code path is heavily optimized so that even a small "if" that only adds one or two assembler instruction has a severe impact on the performance.

joseamrib commented:
> Heureka! It's finally done, full LED control on my Aircube AC.
> 
> It's 3 commits on top of the master/22.03/21.02 branch: https://github.com/stklcode/openwrt/tree/aircube-led-21.02 https://github.com/stklcode/openwrt/tree/aircube-led-22.03 https://github.com/stklcode/openwrt/tree/aircube-led-master (only tested the 22.03 so far)
> 
> The delay has to be enabled in the `make kernel_menuconfig` step: Device Drivers -> SPI support -> Enable delays in the GPIO-based bitbanging SPI Master
> 
> The driver by @c-mauderer which is now upstream works fine. The delay patch does the rest. The LED(-cluster) is dimmable from 0 to 100% with `# echo 0 > /sys/class/leds/ubnt:white:status/brightness` to `# echo 63 > /sys/class/leds/ubnt:white:status/brightness`
> 
> I did some reverse engineering on the hardware. It's the very same Sonix SPI controller, however the GPIO pinout is different from the ISP model (different SoC, why should it be the same)
> 
> pinout image

Hello everyone,

Is it possible to share again the working openwrt builds for the Aircube AC? The links were pulled down. :(.

Thanks in advance.

stklcode commented:
This branch gets updated about every month. I generate patches for my custom images based on main with Kernel 6.1 from there: https://github.com/stklcode/openwrt/tree/main_aircube

It’s just 2 additional commits.

Currently the LED naming is weired, as I tried to replace `label` with `color`/`function` which is not translated correctly to the expected label in the upstream driver. Despite the name, the LED control works fine.

joseamrib commented:
First of all, thanks for the fast reply. I'm sorry, but i'm a little lost here... :( I've installed the openwrt on my Aircube AC using the instructions on the openwrt page, using the file on "Firmware OpenWrt Install". 

It worked hell, but the led is always blinking. :( I'm loving the openwrt firmware because of the SQM (i have a docsis 30/10 net so the aircube it's enough), but the blinking led... :(

Can you please show me a link with a working openwrt firmware, for aircube AC with led control, that i can use Luci to upgrade (sysupgrade)?

stklcode commented:
Official builds do not contain the changes necessary for the SPI LED driver to fully work (see discussion above). You would have to compile the kernel by yourself.

I don’t have any public link for such images. I could publish my custom builds somewhere, but there are quite a few additional changes that will throw errors without proper configuration and I only build the kernel modules I need.
